[Shalini] iP

shaliniseshadri opened the pull request:
# DukePro 

> Your mind is for having ideas, not holding them - David Allen ([source](https://dansilvestre.com/productivity-quotes/)) 

DukePro frees your mind of having to remember things you need to do. It's:

* text-based
* easy to learn
* *SUPER FAST* to use

All you need to do is,

1. download it from [here](https://github.com/shaliniseshadri/ip).
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you 😉

And it is **FREE!**

Features:

 - [ ] Managing tasks
 - [ ] Managing deadlines 
 - [ ] Reminders

Here is the main method: 
```java
public static void main(String[] args) {
        new Duke().run();
    }
```

Commit:
Add Gradle support

shaliniseshadri committed:
Level-1

shaliniseshadri committed:
Level-2

shaliniseshadri committed:
Level-3 A-Classes

shaliniseshadri committed:
Level-4 A-Inheritance

shaliniseshadri committed:
A-TextUiTesting

shaliniseshadri committed:
Level-5

shaliniseshadri committed:
Level-6

shaliniseshadri committed:
Level-6 Formatting Changes and Improved Error Handling

shaliniseshadri committed:
Level-7: Save files

shaliniseshadri committed:
Level-8: Date and Time

shaliniseshadri committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Constants.java
#	src/main/java/Duke.java
#	src/main/java/exceptions/EmptyDescriptionException.java
#	src/main/java/exceptions/IncorrectFormatException.java
#	src/main/java/exceptions/InvalidKeywordException.java
#	src/main/java/exceptions/InvalidNumberInputException.java
#	src/main/java/exceptions/NonExistentTaskException.java
#	src/main/java/exceptions/RepeatedDoneException.java
#	src/main/java/exceptions/RepeatedTaskException.java
#	src/main/java/task/Deadline.java
#	src/main/java/task/Event.java
#	src/main/java/task/ToDo.java

Resolved all conflicts

shaliniseshadri committed:
A-MoreOOP

shaliniseshadri committed:
A-Packages

shaliniseshadri committed:
A-JUnit

shaliniseshadri committed:
A-JUnit - Fixed formatting issues

shaliniseshadri committed:
A-Jar: Package the App as a JAR file + Formatting fixes

shaliniseshadri committed:
A-JavaDoc: Added JavaDocs

shaliniseshadri committed:
A-CodingStandard: Changed parts of the code to comply with the given coding standard

shaliniseshadri committed:
Level-9: Find

shaliniseshadri committed:
Merge branch 'branch-A-CodingStandard'

shaliniseshadri committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/commands/DeleteCommand.java
#	src/main/java/duke/commands/DoneCommand.java
#	src/main/java/duke/commands/TasksOnCommand.java
#	src/main/java/duke/exceptions/DukeException.java
#	src/main/java/duke/parser/Parser.java
#	src/main/java/duke/storage/Storage.java

shaliniseshadri committed:
Fix code styling to comply with coding standards

shaliniseshadri committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

caipng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,61 @@
+    private final String FOLDERPATH = "./data";
+
+    /**
+     * Public constructor which initialises the ui, storage and tasklist before the chatbot runs
```
There are some violations of the [coding standards](https://se-education.org/guides/conventions/java/intermediate.html#statements) with regards to your Javadoc comments, specifically
> In method header comments, the first sentence should start in the form Returns ..., Sends ..., Adds ...

Maybe you can consider something like
```suggestion
     * Constructs a new Duke chatbot, initialising the UI, storage, and task list before running.
```

shaliniseshadri replied:
Hi! Thank you for your comments and suggestions. Will look into them and make the necessary changes!

caipng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,61 @@
+     */
+    public void run() {
+        ui.init();
+        boolean isExit = false;
```
Minor nit: Grammatical error
```suggestion
        boolean shouldExit = false;
```

caipng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,61 @@
+    public static void main(String[] args) {
+        new Duke().run();
+    }
+}
```
Missing new line at end of file.

caipng commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,44 @@
+    /**
+     * Constructor - Initialises the task to be added.
+     *
+     * @param task input task
```
As per the coding standard, how about
```suggestion
     * @param task input Task to be added.
```

caipng commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,44 @@
+        return false;
+    }
+
+    /**
+     * Takes in a list of tasks and the ui. Checks whether the task is already present in the list of tasks
+     * and if it is, it throws a RepeatedTaskException. Else it adds the task to the list of tasks.
+     *
+     * @param taskList the list of tasks upon which the operations need to be performed
+     * @param ui the ui in which the result message of the command execution is displayed to the user
+     */
```
The comments may be slightly verbose. What do you think about
```suggestion
    /**
     * Adds the specified task to the task list if it is not already inside, throws a RepeatedTaskException otherwise.
     *
     * @param taskList Current list of tasks.
     * @param ui Instance of the UI.
     */
```

caipng commented on src/main/java/duke/commands/ListCommand.java:
```diff
@@ -0,0 +1,35 @@
+     * @param taskList the list of tasks upon which the operations need to be performed
+     * @param ui the ui in which a message relating to the command execution is displayed to the user
+     */
+
```
Extra line
```suggestion
```

caipng commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,274 @@
+            String[] parsedString = input.split("\\s", 2);
+            String word = parsedString[1].trim();
+            return new FindCommand(word);
+        } catch (ArrayIndexOutOfBoundsException ex) {
```
I think it is best to avoid using exceptions for control flow.
How about a guard clause like
```java
if (parsedString.length < 2) {
    throw new EmptyDescriptionException("...");
}
```

caipng commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,274 @@
+            int index = Integer.parseInt(indexOfTask);
+            if (index > 0 && index > taskList.size()) {
+                throw new NonExistentTaskException();
+            } else {
+                return new DoneCommand(index);
+            }
```
The `else` may be omitted.
```suggestion
            }
            
            return new DoneCommand(index);
```

caipng commented on src/test/java/duke/taskListTest/TaskListTest.java:
```diff
@@ -0,0 +1,40 @@
+import duke.tasklist.TaskList;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
```
This is disallowed by the coding standards, which specifies that
>  Imported classes should always be listed explicitly

caipng reviewed (commented):
Overall, the coding standards have been mostly adhered to!
However, there are some violations that you may want to look at, mostly regarding the [Javadoc form](https://se-education.org/guides/conventions/java/intermediate.html#comments).

shaliniseshadri committed:
A-Checkstyle: Apply checkstyle to fix  code styling

shaliniseshadri committed:
Add basic GUI

xiangjunn commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,61 @@
+    private final Ui ui;
+    private final Storage storage;
+    private final String FILEPATH = "./data/duke.txt";
+    private final String FOLDERPATH = "./data";
```
Maybe you can use underscore to separate the words?
```suggestion
    private final String FILE_PATH = "./data/duke.txt";
    private final String FOLDER_PATH = "./data";
```

xiangjunn commented on src/main/java/duke/commands/ByeCommand.java:
```diff
@@ -0,0 +1,32 @@
+     * Default constructor for ByeCommand
+     */
+    public ByeCommand() {
+    }
```
I feel that there is no need to include this since there is no other constructors in this file, so the default would already be this. What do you think?

xiangjunn commented on src/main/java/duke/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    /**
+     * Deletes a task from the list provided the index is a valid integer within the range of the list
+     * otherwise throws an InvalidNumberInputException
```
Just a really small thing, but maybe you can add fullstop to your JavaDoc description? Same goes for other methods as well.

xiangjunn commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,44 @@
+            taskList.add(task);
+            ui.displayAdd(task, taskList);
+        }
+    }
```
I like that the code is easy to understand! Good choice of variable and method names 👍

xiangjunn commented on src/main/java/duke/commands/TasksOnCommand.java:
```diff
@@ -0,0 +1,57 @@
+        TaskList list = new TaskList();
+        for (int i = 0; i < taskList.size(); i++) {
+            Task task = taskList.get(i);
+            if (task instanceof Deadline) {
+                if (((Deadline) task).getDate().equals(date)) {
+                    list.add(task);
+                }
+            }
+            if (task instanceof Event) {
+                if (((Event) task).getDate().equals(date)) {
+                    list.add(task);
+                }
+            }
```
Since a task is either a Deadline or Event or any other classes that are subclass of Task, it won't be a Deadline or Event at the same time. Hence, I feel that using "if" and "else if" is better than "if and if" since you only need to check if a task is an instance of Event only if it is not an instance of Deadline.
```suggestion
            if (task instanceof Deadline) {
                if (((Deadline) task).getDate().equals(date)) {
                    list.add(task);
                }
            } else if (task instanceof Event) {
                if (((Event) task).getDate().equals(date)) {
                    list.add(task);
                }
            }
```

xiangjunn commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,273 @@
+        String[] parsedInput = input.trim().split("\\s", 2);
+        String command = parsedInput[0];
+
+        switch(command.toUpperCase()) {
+            case "BYE":
```
There should be no indentation for case clauses according to Java coding standard.

xiangjunn commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,146 @@
+                }
+                break;
+            case "D":
+                task = new Deadline(parsed[2].trim(), DateTimeParser.readDate(parsed[3].trim()));
```
Maybe can break this line into more lines i.e do parsed[2].trim() and DateTimeParser.readDate(parsed[3].trim() on separate lines for more readability.

xiangjunn commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,146 @@
+        String str;
+
+            switch (taskType) {
+            case "[T]":
+                str = "T" + " | " + task.getIntStatus() + " | " + task.getDescription() + "\n";
+                break;
+            case "[D]":
+                str = "D" + " | " + task.getIntStatus() + " | " + task.getDescription() + " | "
+                        + task.getDateString() + "\n";
+                break;
+            case "[E]":
+                str = "E" + " | " + task.getIntStatus() + " | " + task.getDescription() + " | "
+                        + task.getDateString() + "\n";
```
Can consider using string.format for more readability?

xiangjunn commented on src/test/java/duke/taskListTest/TaskListTest.java:
```diff
@@ -0,0 +1,44 @@
+
+
+    @Test
+    public void getTaskTest() {
+        loadTestList(testList);
+        int index = 1;
+        assertEquals(new ToDo("clean room").toString(), testList.get(index - 1).toString());
+    }
```
I feel that this is not tested in isolation since bugs for ToDo/Deadline/Event can affect this test case.

xiangjunn reviewed (commented):
LGTM. Just some minor issues here and there. Overall, the quality of code is great! 😄

shaliniseshadri committed:
Improve UI Design

shaliniseshadri committed:
Generate a new JAR file

shaliniseshadri committed:
Add assertions

shaliniseshadri committed:
A-CodeQuality - Javadocs formatting

The current code had quite a few issues with the Javadocs format.

The javadocs format beginning with simple present tense (Returns, Encapsulates, Parses) is more beneficial to those reading the code as it helps make the function of the method/class clearer.

Let's,
* make the according changes so that the javadocs begin with the above-mentioned words.

shaliniseshadri committed:
A-CodeQuality: Remove all unused methods and import statements and fix Javadocs formatting

The current code has some unused statements and methods as a result of refactoring and some of the Javadocs had were not of the correct format

Removing all unused methods and using clearer javadocs would prevent confusions and make the code clearer for those reading it

Let's
*remove all unused methods and import statements
*reframe the incorrectly formatted javadocs

shaliniseshadri committed:
Merge pull request #1 from shaliniseshadri/branch-A-CodeQuality

branch-A-CodeQuality

shaliniseshadri committed:
Merge pull request #2 from shaliniseshadri/branch-A-Assertions

Add assertions

shaliniseshadri committed:
C-Update: Add basic task editing functionality

shaliniseshadri committed:
Improve UI

shaliniseshadri committed:
Add README

shaliniseshadri committed:
Fix coding standard and add UI screenshot

shaliniseshadri committed:
Update product screenshot

shaliniseshadri committed:
Rename product screenshot

shaliniseshadri committed:
Fix error message description and empty description exception

shaliniseshadri committed:
Fix font formatting issue and warning while exporting to jar

shaliniseshadri committed:
Update product screenshot
